push-ing pragma exportc generates invalid C identifiers

### Description

Compiles fine

```nim
type r* = object

proc foo(): r {.exportc: "$1".} =
  echo result
```

Fails:

```nim
type r* = object

{.push exportc: "$1".}
proc foo(): r =
  echo result
{.pop.}
```

Compilation command:

```bash
nim -r c repro.nim
```

### Nim Version

```
Nim Compiler Version 2.0.0 [MacOSX: arm64]
Compiled at 2023-08-02
Copyright (c) 2006-2023 by Andreas Rumpf

active boot switches: -d:release
```

### Current Output

Fails with an invalid generated C symbol:

```text
CC: ../../nim-2.0.0/lib/std/objectdollar.nim
/Users/turbo/.cache/nim/repro_d/@m..@s..@snim-2.0.0@slib@sstd@sobjectdollar.nim.c:142:10: error: expected ';' at end of declaration
        NI count`gensym1;
                ^
                ;
/Users/turbo/.cache/nim/repro_d/@m..@s..@snim-2.0.0@slib@sstd@sobjectdollar.nim.c:149:7: error: expected ';' after expression
        count`gensym1 = ((NI)0);
             ^
             ;
/Users/turbo/.cache/nim/repro_d/@m..@s..@snim-2.0.0@slib@sstd@sobjectdollar.nim.c:149:7: error: expected expression
```

### Expected Output

none, successful compilation

### Possible Solution

_No response_

### Additional Information

_No response_

--- a/compiler/pragmas.nim
+++ b/compiler/pragmas.nim
@@ -1313,7 +1313,7 @@ proc implicitPragmas*(c: PContext, sym: PSym, info: TLineInfo,
   if sym != nil and sym.kind != skModule:
     for it in c.optionStack:
       let o = it.otherPragmas
-      if not o.isNil and sfFromGeneric notin sym.flags: # see issue #12985
+      if not o.isNil: # see issue #12985
         pushInfoContext(c.config, info)
         var i = 0
         while i < o.len:

--- a/compiler/pragmas.nim
+++ b/compiler/pragmas.nim
@@ -1313,7 +1313,7 @@ proc implicitPragmas*(c: PContext, sym: PSym, info: TLineInfo,
   if sym != nil and sym.kind != skModule:
     for it in c.optionStack:
       let o = it.otherPragmas
-      if not o.isNil: # see issue #12985
+      if not o.isNil:
         pushInfoContext(c.config, info)
         var i = 0
         while i < o.len:

--- a/compiler/seminst.nim
+++ b/compiler/seminst.nim
@@ -400,11 +400,17 @@ proc generateInstance(c: PContext, fn: PSym, pt: TIdTable,
     entry.compilesId = c.compilesContextId
     addToGenericProcCache(c, fn, entry)
     c.generics.add(makeInstPair(fn, entry))
+    # bug #12985 bug #22913
+    # TODO: use the context of the declaration of generic functions instead
+    # TODO: consider fixing options as well
+    let otherPragmas = c.optionStack[^1].otherPragmas
+    c.optionStack[^1].otherPragmas = nil
     if n[pragmasPos].kind != nkEmpty:
       pragma(c, result, n[pragmasPos], allRoutinePragmas)
     if isNil(n[bodyPos]):
       n[bodyPos] = copyTree(getBody(c.graph, fn))
     instantiateBody(c, n, fn.typ.n, result, fn)
+    c.optionStack[^1].otherPragmas = otherPragmas
     sideEffectsCheck(c, result)
     if result.magic notin {mSlice, mTypeOf}:
       # 'toOpenArray' is special and it is allowed to return 'openArray':

--- a/tests/pragmas/tpush.nim
+++ b/tests/pragmas/tpush.nim
@@ -38,3 +38,42 @@ proc main(): void =
   {.push staticBoundChecks: on.}
 
 main()
+
+
+proc timnFoo[T](obj: T) {.noSideEffect.} = discard # BUG
+
+{.push exportc.}
+proc foo1() =
+  var s1 = "bar"
+  timnFoo(s1)
+  var s2 = @[1]
+  timnFoo(s2)
+{.pop.}
+
+
+block: # bug #22913
+  block:
+    type r = object
+
+    template std[T](x: T) =
+      let ttt {.used.} = x
+      result = $ttt
+
+    proc bar[T](x: T): string =
+      std(x)
+
+    {.push exportc: "$1".}
+    proc foo(): r =
+      let s = bar(123)
+    {.pop.}
+
+    discard foo()
+
+  block:
+    type r = object
+    {.push exportc: "$1".}
+    proc foo2(): r =
+      let s = $result
+    {.pop.}
+
+    discard foo2()

